File: backend/app/schemas/application_event.py
from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict


class ApplicationTransitionRequest(BaseModel):
    target_stage: str
    next_action: str | None = None
    next_action_at: datetime | None = None
    notes: str | None = None
    event_at: datetime | None = None
    operator_type: str = "user"
    payload_json: dict[str, Any] | None = None
    note: str | None = None


class ApplicationEventListItem(BaseModel):
    id: int
    application_record_id: int
    event_type: str
    from_stage: str | None
    to_stage: str | None
    event_at: datetime | None
    operator_type: str
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class ApplicationEventRead(ApplicationEventListItem):
    payload_json: dict[str, Any] | None
    note: str | None

File: frontend/src/types/application_event.ts
import type { ISODateString, JsonObject } from "./common";

export interface ApplicationTransitionRequest {
  target_stage: string;
  next_action?: string | null;
  next_action_at?: ISODateString | null;
  notes?: string | null;
  event_at?: ISODateString | null;
  operator_type?: string;
  payload_json?: JsonObject | null;
  note?: string | null;
}

export interface ApplicationEventListItem {
  id: number;
  application_record_id: number;
  event_type: string;
  from_stage: string | null;
  to_stage: string | null;
  event_at: ISODateString | null;
  operator_type: string;
  created_at: ISODateString;
}

export interface ApplicationEvent extends ApplicationEventListItem {
  payload_json: JsonObject | null;
  note: string | null;
}
